Source/Metal/CommandBufferMTL.mm

```diff
@@ -0,0 +1,689 @@
+    m_CurrentPipelineLayout = &pipelineLayoutMTL;
+}
+
+void CommandBufferMTL::SetDescriptorSet(uint32_t setIndexInPipelineLayout, const DescriptorSet& descriptorSet, const uint32_t* dynamicConstantBufferOffsets) {
+    const DescriptorSetMTL& descriptorSetImpl = (DescriptorSetMTL&)descriptorSet;
+    auto* layout = m_CurrentPipelineLayout->GetDescriptorSetLayout(setIndexInPipelineLayout);
+    const uint32_t space = layout->m_DescriptorSetDesc.registerSpace;
+    
+    StageBits stageBits = (StageBits)0;
+    for(size_t i = 0; i < layout->m_DescriptorSetDesc.rangeNum; i++) {
+        if(layout->m_DescriptorSetDesc.ranges[i].shaderStages == StageBits::NONE)
+            continue;
+        stageBits |= layout->m_DescriptorSetDesc.ranges[i].shaderStages ;
+    }
+    if(stageBits == StageBits::ALL ||
+       stageBits & StageBits::FRAGMENT_SHADER) {
+        [m_RendererEncoder setFragmentBuffer: descriptorSetImpl.GetArgumentBuffer()
+                                      offset: descriptorSetImpl.GetArugmentBufferOffset()
+                                     atIndex: space];
+    }
+    if(stageBits == StageBits::ALL ||
+       stageBits & StageBits::VERTEX_SHADER) {
+        [m_RendererEncoder setVertexBuffer: descriptorSetImpl.GetArgumentBuffer()
+                                    offset: descriptorSetImpl.GetArugmentBufferOffset()
+                                   atIndex: space];
+        
+    }
+    
+    if(stageBits == StageBits::ALL ||
+       stageBits & StageBits::COMPUTE_SHADER) {
+        [m_ComputeEncoder setBuffer: descriptorSetImpl.GetArgumentBuffer()
+                                    offset: descriptorSetImpl.GetArugmentBufferOffset()
+                                   atIndex: space];
+        
+    }
+    
+}
+
+void CommandBufferMTL::SetConstants(uint32_t pushConstantIndex, const void* data, uint32_t size) {
+    const struct RootConstantDesc* pDesc = m_CurrentPipelineLayout->GetPushBinding(pushConstantIndex);
+    
+    
+    if (pDesc->shaderStages == StageBits::ALL ||
+        pDesc->shaderStages & StageBits::VERTEX_SHADER)
+    {
+        [m_RendererEncoder setVertexBytes:data length:size atIndex:pDesc->registerIndex];
+    }
+    
+    if (pDesc->shaderStages == StageBits::ALL ||
+        pDesc->shaderStages & StageBits::FRAGMENT_SHADER )
+    {
+        [m_RendererEncoder setFragmentBytes:data length:size atIndex:pDesc->registerIndex];
+    }
+    
+    if (pDesc->shaderStages == StageBits::ALL ||
+        pDesc->shaderStages & StageBits::COMPUTE_SHADER )
+    {
+        [m_ComputeEncoder setBytes:data length:size atIndex:pDesc->registerIndex];
+    }
+}
+
+void CommandBufferMTL::SetDescriptorPool(const DescriptorPool& descriptorPool) {
+    
+}
+
+
+void CommandBufferMTL::InsertBarriers() {
+    if(m_RendererEncoder) {
+        if (m_barrierFlags & BarrierBits::BARRIER_FLAG_BUFFERS)
+        {
+            [m_RendererEncoder memoryBarrierWithScope:MTLBarrierScopeBuffers
```

Comment:
I don't think this is sufficient at all.

Metal memory barriers only impact the pass they are in: https://developer.apple.com/videos/play/wwdc2022/10101/?time=1495
So it could still overlap with a different pass in the same command buffer or with another command buffer.

Meanwhile Vulkan barriers (and I assume the NRI barriers that look very similar) go beyond that and impact the whole queue.